pkg/kube/object_patch/operation.go

```diff
@@ -0,0 +1,253 @@
+	ignoreMissingObject bool
+
+	// Delete options.
+	deletionPropagation metav1.DeletionPropagation
+}
+
+func (op *Operation) applyOptions(options ...OperationOption) {
+	for _, option := range options {
+		option(op)
+	}
+}
+
+func (op *Operation) isCreate() bool {
+	return op.object != nil
+}
+
+func (op *Operation) isDelete() bool {
+	return op.deletionPropagation != ""
+}
+
+func (op *Operation) isPatch() bool {
+	return op.patchType != ""
+}
+
+func (op *Operation) isFilter() bool {
+	return op.filterFunc != nil
+}
+
+func (op *Operation) Description() string {
+	if op.isCreate() {
+		return "Create object"
+	}
+	if op.isDelete() {
+		return fmt.Sprintf("Delete object %s/%s/%s/%s", op.apiVersion, op.kind, op.namespace, op.name)
+	}
+	if op.isPatch() {
+		return fmt.Sprintf("Patch object %s/%s/%s/%s using %s patch", op.apiVersion, op.kind, op.namespace, op.name, op.patchType)
+	}
+	if op.isFilter() {
+		return fmt.Sprintf("Filter object %s/%s/%s/%s", op.apiVersion, op.kind, op.namespace, op.name)
+	}
+	return "Unknown operation"
+}
+
+//		operationError = o.Create(&unstructured.Unstructured{Object: spec.Object},
+
+func NewFromOperationSpec(spec OperationSpec) *Operation {
+	switch spec.Operation {
+	case Create:
+		return NewCreateOperation(spec.Object,
+			WithSubresource(spec.Subresource))
+	case CreateIfNotExists:
+		return NewCreateOperation(spec.Object,
+			WithSubresource(spec.Subresource),
+			IgnoreIfExists())
+	case CreateOrUpdate:
+		return NewCreateOperation(spec.Object,
+			WithSubresource(spec.Subresource),
+			UpdateIfExists())
+	case Delete:
+		return NewDeleteOperation(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource))
+	case DeleteInBackground:
+		return NewDeleteOperation(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			InBackground())
+	case DeleteNonCascading:
+		return NewDeleteOperation(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			NonCascading())
+	case JQPatch:
+		return NewFilterPatchOperation(func(u *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+			return applyJQPatch(spec.JQFilter, u)
+		},
+			spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		)
+	case MergePatch:
+		return NewMergePatchOperation(spec.MergePatch,
+			spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+			WithSubresource(spec.Subresource),
+			WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		)
+		//mergePatch, err := json.Marshal(spec.MergePatch)
+		//if err != nil {
+		//	applyErrors = multierror.Append(applyErrors, err)
+		//	continue
+		//}
+		//operationError = o.Patch(spec.ApiVersion, spec.Kind, spec.Namespace, spec.Name,
+		//	WithSubresource(spec.Subresource),
+		//	UseMergePatch(mergePatch),
+		//	WithIgnoreMissingObject(spec.IgnoreMissingObject),
+		//)
```

Comment:
PatchCollector was able to return an error if the patch is not valid. Now, PatchCollector saves the patch as-is. Patch content is handled by the ObjectPatcher during ExecuteOperations after hook execution.